pyrdp/logging/StatCounter.py

```diff
@@ -0,0 +1,158 @@
+
+    DEVICE_REDIRECTION_IOERROR = "deviceRedirectionIOError"
+    # IO error packets for the rdpdr channel
+
+    DEVICE_REDIRECTION_FILE_CLOSE = "deviceRedirectionFileClose"
+    # File Close packets for the rdpdr channel
+
+    DEVICE_REDIRECTION_FORGED_FILE_READ = "deviceRedirectionForgedFileRead"
+    # File read packets forged by pyrdp for the rdpdr channel
+
+    DEVICE_REDIRECTION_FORGED_DIRECTORY_LISTING = "deviceRedirectionForgedDirectoryListing"
+    # Directory listing packets forged by pyrdp for the rdpdr channel
+
+    CLIPBOARD = "clipboard"
+    # Number of clipboard PDUs coming from either end
+
+    CLIPBOARD_CLIENT = "clipboardClient"
+    # Number of clipboard PDUs coming from the client
+
+    CLIPBOARD_SERVER = "clipboardServer"
+    # Number of clipboard PDUs coming from the server
+
+    CLIPBOARD_COPY = "clipboardCopies"
+    # Number of times data has been copied by either end
+
+    CLIPBOARD_PASTE = "clipboardPastes"
+    # Number of times data has been pasted by either end
+
+
+class StatCounter:
+    """
+    Class that keeps track of various statistics during an RDP connection (See STAT)
+    """
+
+    def __init__(self):
+        self.stats = {"report": 1.0}  # 1.0 = True
+
+    def increment(self, *args: str):
+        """
+        Increments all statistics passed in arguments
+        :param args: list of statistics to increment by one. See STAT for list of allowed values.
+        """
+        for stat in args:
+            if stat not in self.stats:
+                self.stats[stat] = 0
+            self.stats[stat] += 1
+
+    def incrementWith(self, statDestination: str, *statsSource: str):
+        """
+        Increments statDestination by all provided statSources
+        """
+        if statDestination not in self.stats:
+            self.stats[statDestination] = 0
+        for statSource in statsSource:
+            if statSource in self.stats:
+                self.stats[statDestination] += self.stats[statSource]
+
+    def start(self):
+        """
+        Initialize some statistics such as connectionTime
+        """
+        self.stats[STAT.CONNECTION_TIME] = time.time()
+
+    def stop(self):
+        """
+        Calculates the last statistics such as interaction ratio and connectionTime
+        """
+        self.stats[STAT.CONNECTION_TIME] = time.time() - self.stats[STAT.CONNECTION_TIME]
+        self.incrementWith(STAT.TOTAL_INPUT, STAT.MCS_INPUT, STAT.IO_INPUT_FASTPATH, STAT.VIRTUAL_CHANNEL_INPUT, STAT.CLIPBOARD_CLIENT, STAT.DEVICE_REDIRECTION_CLIENT)
+        self.incrementWith(STAT.TOTAL_OUTPUT, STAT.MCS_OUTPUT, STAT.IO_OUTPUT_FASTPATH, STAT.VIRTUAL_CHANNEL_OUTPUT, STAT.CLIPBOARD_SERVER, STAT.DEVICE_REDIRECTION_SERVER)
+        if self.stats[STAT.TOTAL_OUTPUT] > 0:
+            self.stats[STAT.CLIENT_SERVER_RATIO] = self.stats[STAT.TOTAL_INPUT] / self.stats[STAT.TOTAL_OUTPUT]
+
+    def logReport(self, log: LoggerAdapter):
+        log.info(self.stats)
```

Comment:
The log report doesn't respect the JSON format of every other logger, so the stats can't be indexed like every other field.

Reply:
Fixed in last commit